### ISSUE TITLE
API does not prompt for basic auth credentials over https

### ISSUE THREAD

**Issue by judgej** [woothemes/woocommerce#8256]
I have been using the API from a PHP library and also directly from a browser. The “Consumer key is missing” error happens all too often, but the explanation of this being a "server fault" most of the time does not quite ring true. Here's why.

I have tried accessing the API direct from a browser:

e.g. https://ewxample.com/wc-api/v2/products

That gives me the “Consumer key is missing” error. But why isn't the browser asking me for the credentials? A few lines to return the right headers instead of an error, and it does, as detailed here:

http://php.net/manual/en/features.http-auth.php

Credentials can be entered, and the GET functions of the API can be accessed through the browser. That's great, and really helps to understand how the API works, as well as following the basic auth spec a little more closely.

On the same server I was having troubles accessing the same API over SSL from a PHP library and had to fall back to non-SSL oauth method. I had assumed the issue was the server config, but if a browser can get in this way, it can't be the server at fault, so I'm guessing there is something else going on here. I'm still investigating that.

Anyway, tl;dr bit: accessing the API using SSL, which requires Basic Realm auth, should return the appropriate headers to prompt for the credentials, and not a JSON error message. I can provide a PR for this if the explanation is acceptable.

**Comment by judgej** [comment]
A "logout" entry point may also be useful if this is implemented, since the browser will be "logged in" to basic auth until it is totally closed or a forced logout is done (just change the realm to force this).

**Comment by judgej** [comment]
The "Consumer Key is missing" error in perform_ssl_authentication() returns a 404 too. Should that not be a 401 anyway?

**Comment by judgej** [comment]
Just to expand on this a little, the authentication logic in `perform_ssl_authentication()` effectively takes one of two paths:

* If the `consumer_secret` GET parameter is provided, then a kind of "fake" Basic auth is implemented. The consumer token comes from the request header, and the consumer key comes from the GET parameter. Any authentication errors should result in a JSON response, as they do now, because this is not Basic auth (so far as I can see), but something that kind of uses some of the Basic auth features, e.g. expecting PHP to pull the appropriate header into the PHPO_AUTH_USER sugerglobal.
* If the `consumer_secret` GET parameter is NOT provided, then we have full-on Basic auth. Any validation errors should result in the appropriate headers being returned to indicate that Basic auth is in effect at that endpoint (e.g. 'WWW-Authenticate: Basic realm="WooCommerce API"' and a 'HTTP/1.0 401 Unauthorized').

This will allow browsers to access the API in the standards-compliant way that browsers do, and also other services to access the API using Basic auth and handle authentication errors in an expected way. It also retains the fake-Basic auth workaround that is in place for whatever reason it is there for, so retaining backward-compatibility.

**Comment by claudiosmweb** [comment]
@maxrice can help me here?

**Comment by judgej** [comment]
I've been trying a few things on out my test system, and reached the conclusion that the SSL auth should be handling either the GET parameters as a custom auth method, or HTML Basic Auth, but not a hybrid of the two. So if the key/secret GET parameters are there, then use them and use them only. If they are not, then fall back to Basic Auth and implement that protocol fully, including the appropriate 401 headers.

At the moment, you can add the GET parameters, but if the browser happens to have Basic Auth user and/or password set, then that will override the GET parameters. For example if you go to as https://xxx@example.com/wc-api/v2/whatever then that breaks any consequent attempts to connect using ?consumer_key=user&consumer_secret=password, because the browser remembers the "xxx" and sends that along with the request.

**Comment by maxrice** [comment]
@judgej very nice write-up of the issues here. You're correct in that currently it's a sort of odd combination of both quasi-standard HTTP Basic Auth that falls back to a custom query string auth when accessing the API over SSL. The history here is that originally there was no fallback for Basic Auth, and what we found was that roughly 10-15% of servers couldn't provide `PHP_AUTH_USER` or `PHP_AUTH_PW`. This mostly occurs with IIS server (though I've seen some Apache servers mangle the auth header as well).

While it's possible to fix this via the IIS config, it was unlikely that a user would wanted to simply use the WC iPhone app would go through the trouble of contacting their host, etc. However, we still preferred to use SSL where available, so we implemented a query string auth fallback in #5277 similar to how Stripe allowed query string auth over SSL.

I agree with your suggestions to only use the query string auth if the parameters are present, and otherwise fall back to standards-compliant Basic Auth -- it'd be awesome to have you work up a PR :smile: I'm happy to help out too.

@claudiosmweb I don't expect any backwards-incompatible changes as part of this, but definitely best to put it in WC 2.4 and bump the major version to `v3`.

**Comment by judgej** [comment]
Cool. I'll work on it over the week :-)

I've seen the PHP_AUTH_* elements missing on Apache for FastCGI. I'm not sure if there is a workaround for that to implement Basic Auth (maybe the underlying headers can be read instead). I'll explore that.

Been reading up on Digest Auth too, which may be fun to implement if I was twiggling my fingers, which is unlikely to every happen ;-) I see there is a filter for auth, so in theory any alternative method of authenticating the API user can be plugged in if there is a need, so it's nicely future-proofed anyway.

So when writing a PR, is it best to reference it here and add notes to this ticket as I go along? Or keep any notes on the PR and link that to this ticket when its ready to be tested?

**Comment by maxrice** [comment]
Yeah looking over my old notes, it seems like Apache under FCGI doesn't always pass the headers properly either. It's fairly challenging to build a cross-platform method in PHP that reads the raw HTTP headers which is one reason why the OAuth implementation uses query strings instead of the Authorization header :person_frowning: 

I think you could definitely build a plugin to support digest auth, though IMO it's not a very common method (currently) for API auth, at least when you look at the majors (Stripe, GitHub, etc)

I always like to see notes on the PR, rather than the issue as it's easier to reference and comment on that when reviewing the code :) I also usually add a comment to the issue like "will be fixed in #{PR issue number}" so it's cross-referenced then too.

**Comment by judgej** [comment]
Still on my list - just scheduling it in...

**Comment by mikejolley** [comment]
@judgej still planning on tackling this? If not @claudiosmweb or @justinshreve can take a look :)

**Comment by judgej** [comment]
Still planning to fit it in, but don't let me hold you back if you want to go ahead. I don't have any code for it at this stage (apart from a proof-of-concept `echo` stuffed in the API code) so we are all at the same stage.

**Comment by justinshreve** [comment]
@judgej I'm taking a look at this in the following branch https://github.com/justinshreve/woocommerce/tree/api-basic-auth-8256 -- nothing to look at yet (only changes so far is versioning the files per the discussion above) but wanted to give you a heads up. I'll ping you when I have something to review :).

**Comment by justinshreve** [comment]
@judgej a PR was submitted in #8406 -- would appreciate your feedback. Specifically the `includes/api/class-wc-api-authentication.php` file.

**Comment by judgej** [comment]
Giving this a go now. I see the token generation has been moved to the WC settings pages.

It took a few goes logging in, before it worked. I revoked the key, regenerated another, then went to /wc-api/v3/reauth - that behaves strangely, because it asks for the login credentials, and immediately invalidates them because you go back to reauth on entering your details. It should perhaps redirect to the index page /wc-api/v3 upon successfully invalidating the login credentials (however it does that - I understand there is no one way that will work for all browsers with BASIC auth).

Once I realised I was stuff in a loop there, I went to /wc-auth/v3/products, entered my auth details, and that worked. I'll run a few more tests and see how it goes.

**Comment by judgej** [comment]
Upon cancelling a login credentials pop-up in the browser, a text message informing me that the API needs me to be logged in is returned. However, it is plain text, while the page header lists it as JASON:

Content-Type: application/json; charset=UTF-8

If the page is being delivered as JSON, then the contents should probably be JSON too, always, so the text error message should be wrapped, i.e. JSON encoded. There is a JSON error message format already used in some of the other API calls, for example:

{"errors":[{"code":"woocommerce_api_no_product_found","message":"No product found with the ID equal to 123"}]}

**Comment by judgej** [comment]
On /wc-api/v3/orders I am getting deprecation errors:

Notice: get_formatted_legacy is <strong>deprecated</strong> since version 2.4! Use Item Meta Data is being called with legacy arguments instead. in /var/www/vhosts/acadweb.co.uk/httpdocs/printtrail-shop/wp-includes/functions.php on line 3391

It looks like this is being generated in ./includes/class-wc-order-item-meta.php I'll check in case it is a plugin I have installed doing that, but that's a heads-up in case it is something easy to fix (and it's not me).

**Comment by justinshreve** [comment]
Good point about /reauth. I'll add a redirect.

I'll fix up the error as well - thanks!

For the deprecation notice, I get it too, but it's unrelated to this PR/issue. Something in the /orders endpoint is calling a deprecated function. That should be fixed, but we should open a separate PR/issue for it.

I'll make some commits here in a few minutes to fix the first two issues. Thank you!

**Comment by judgej** [comment]
Thanks! I'll switch off debug and see how it works with my app that scans for new orders. I do have custom metadata added to order lines. I link each product order line to the shipping order line in the checkout process and vice-versa, as my app needs to know how each item in the order is going to be shipped. I was just a little worried that a change may break that - custom metadata added to shipping lines is delivered through the API V2 nicely.

**Comment by judgej** [comment]
Yes, the custom order line meta fields are still delivered. But now so are the private meta fields - _qty _line_tax etc. I've just gone back to v2, and there are there too. I'm sure they weren't previously (which is why I made sure my order-line linking fields were not "_" prefixed to hide them. It's not causing a problem, but may be a side-effect of the changes that you don't want.

**Comment by justinshreve** [comment]
I'm actually thinking that /reauth is not going to work the way we want it to (or a logout, either). Redirecting isn't going to work. I think we should drop that idea. Searching around, it doesn't really seem like there are great solutions - so we can just document that the session needs to be cleared by the client (wether that is from clearing browser sessions or otherwise).

For differences in API output, that should also be a separate PR/issue if something doesn't look right. This branch is just a direct copy of what is already in master, so no output is changed in this PR.

I committed a fix for the JSON error.

**Comment by judgej** [comment]
Okay. I'm in two minds to raise the ticket about the meta fields, as I'm not sure whether it is an error or not. I might do it anyway for visibility.

The reauth thing - yeah - I did some research and also could not find a consistent way to make it work. Digect Auth has a mechanism built in, but Basic simply hasn't. So practically, clearing it from the browser end is probably be safest thing to do. We always have the GET parameters to fall back to if need be, that should override the Basic Auth header details.

**Comment by judgej** [comment]
Error on cancelling the auth pop-up looks good now (JSON).

**Comment by justinshreve** [comment]
Yup, agreed. and with these changes it does override the header details, so we are good there :).

**Comment by judgej** [comment]
Going to this URL (putting a dummy user in front) will clear the auth. It effectively changes the saved consumer key to anythinghere, so you are forced to log in again. It's a handy trick, and works properly now with your changes:

https://anythinghere@example.co.uk/wc-api/v3/orders
